library/std/src/sys/sync/rwlock/futex.rs

```diff
@@ -54,6 +55,19 @@ fn is_read_lockable(state: u32) -> bool {
     state & MASK < MAX_READERS && !has_readers_waiting(state) && !has_writers_waiting(state)
 }
 
+#[inline]
+fn is_read_lockable_after_wakeup(state: u32) -> bool {
+    // We make a special case for checking if we can read-lock _after_ a reader thread that went to
+    // sleep has been woken up by a call to `downgrade`.
+    //
+    // `downgrade` will wake up all readers and place the lock in read mode. Thus, there should be
+    // no readers waiting and the lock should be read-locked (not write-locked or unlocked).
+    state & MASK < MAX_READERS
+        && !has_readers_waiting(state)
+        && !is_write_locked(state)
+        && !is_unlocked(state)
+}
```

Comment:
Is the assumption here that `has_writers_waiting(state))` is always false since a reader threads are only woken up if there are no writers?

Reply:
So this goes back to the bullet point task at the very top about performance.

Here I am changing the protocol such that if `downgrade` is called, we wake up all readers even if there are writers waiting. The reasoning behind this is that if someone calls `downgrade` on the `RwLock`, then it is reasonable to say that they want other people to read the value they are protecting. Otherwise what was the point of downgrading in the first place? It doesn't make too much sense to say that if there are any writers waiting, `RwLock` `downgrade` only allows the writer who just downgraded to read.

In other words, this is the one case where readers would intentionally take the lock before writers. This additionally provides a mechanism to prefer readers over writers if people want that, but that is a secondary feature.

Reply:
Oh that makes total sense, you wouldn't downgrade a guard just to immediately drop it. Thanks for the explanation, upgrading the docs would be great.